Bukkit/src/main/java/com/plotsquared/bukkit/util/BukkitRegionManager.java

```diff
@@ -88,73 +84,29 @@ public boolean handleClear(
 
     @Override
     public int[] countEntities(@NonNull Plot plot) {
+        int[] count = new int[6];
         int[] existing = (int[]) plot.getMeta("EntityCount");
         if (existing != null && (System.currentTimeMillis() - (long) plot.getMeta("EntityCountTime") < 1000)) {
             return existing;
         }
         PlotArea area = plot.getArea();
         World world = BukkitUtil.getWorld(area.getWorldName());
-        Location bot = plot.getBottomAbs();
-        Location top = plot.getTopAbs();
-        int bx = bot.getX() >> 4;
-        int bz = bot.getZ() >> 4;
-
-        int tx = top.getX() >> 4;
-        int tz = top.getZ() >> 4;
-
-        int size = tx - bx << 4;
-
-        Set<Chunk> chunks = new HashSet<>();
-        for (int X = bx; X <= tx; X++) {
-            for (int Z = bz; Z <= tz; Z++) {
-                if (world.isChunkLoaded(X, Z)) {
-                    chunks.add(world.getChunkAt(X, Z));
-                }
-            }
-        }
-
-        boolean doWhole = false;
-        List<Entity> entities = null;
-        if (size > 200 && chunks.size() > 200) {
-            entities = world.getEntities();
-            if (entities.size() < 16 + size / 8) {
-                doWhole = true;
-            }
+        if(world == null) {
+            return count;
         }
 
-        int[] count = new int[6];
-        if (doWhole) {
-            for (Entity entity : entities) {
-                org.bukkit.Location location = entity.getLocation();
-                PaperLib.getChunkAtAsync(location).thenAccept(chunk -> {
-                    if (chunks.contains(chunk)) {
-                        int X = chunk.getX();
-                        int Z = chunk.getZ();
-                        if (X > bx && X < tx && Z > bz && Z < tz) {
-                            count(count, entity);
-                        } else {
-                            Plot other = area.getPlot(BukkitUtil.adapt(location));
-                            if (plot.equals(other)) {
-                                count(count, entity);
+        for (final CuboidRegion region : plot.getRegions()) {
+            for (int x = region.getMinimumPoint().getX() >> 4; x <= region.getMaximumPoint().getX() >> 4; x++) {
+                for (int z = region.getMinimumPoint().getZ() >> 4; z <= region.getMaximumPoint().getZ() >> 4; z++) {
+                    world.getChunkAtAsync(x, z).thenAccept(chunk -> {
```

Comment:
loading chunks async will break things; `count` is returned before the chunk is loaded (that was flawed before too, I guess). For very large plots, this might be important, but that means we either need to come up with something clever or load all chunks before starting counting.

Reply:
I thought about it and played around, but unfortunately I can't really think of a simple idea that would allow for performance-friendly loading for larger plots without making any really big changes.

Reply:
I think the short-term best solution is to just ignore unloaded chunks. That will result in the same behavior as the current code if I didn't miss anything. It also isn't a problem for a vast majority of servers I guess.

Long-term, we probably need to rethink the concept or come up with a solution to store the entity count of unloaded chunks somehow.

Removing entities later likely is a bad idea, as we would need to track how the entity was created (i.e. via spawn eggs, you'd need to give the spawn egg back to the player, etc...).

Reply:
Thanks for the feedback. As an alternative, I'm also thinking of caching the whole thing somehow.

Because the solution with checkAsync works, but as you already said, too many additional checks would have to be made and the events would not really be cancelable.

The whole thing is breaking my head. If anyone wants to give that, I would be grateful. Otherwise, I would try it again as soon as I have time.

Reply:
You could probably just open an issue for it so we can keep track of it. I don't think it's worth to spend too much time on it right now, the fix for merged plots is far more important.

Reply:
The problem is I had already made several tickets about it, but @OneLiteFeather kept marking them as invalid because she didn't understand the problem and wasn't open to discussing it.

Reply:
We are open for discussions but not mixing up contexts or creating incorrect issues. Also we do no private support including voicechat and discord dms.

Reply:
Anyway, you can now be sure that this problem exists and that you can reproduce it without any problems. I'll write another proper ticket in a few days to record it.

Reply:
I created an issue to keep track of it: #4362